src/mpl/include/mpl_thread_argobots.h

```diff
@@ -52,9 +52,42 @@ typedef void (*MPL_thread_func_t) (void *data);
 void MPL_thread_create(MPL_thread_func_t func, void *data, MPL_thread_id_t * idp, int *errp);
 
 #define MPL_thread_exit()
-#define MPL_thread_self(idp_) ABT_thread_self(idp_)
-#define MPL_thread_join(id_) ABT_thread_free(&id_)
-#define MPL_thread_same(idp1_, idp2_, same_)  ABT_thread_equal(*idp1_, *idp2_, same_)
+#define MPL_thread_self(idp_)                                                 \
+    do {                                                                      \
+        ABT_thread self_thread_tmp_ = ABT_THREAD_NULL;                        \
+        ABT_thread_self(&self_thread_tmp_);                                   \
+        uintptr_t id_tmp_;                                                    \
+        if (self_thread_tmp_ == ABT_THREAD_NULL) {                            \
+            /* It seems that an external thread calls this function. */       \
+            /* Use Pthreads ID instead. */                                    \
+            id_tmp_ = (uintptr_t)pthread_self();                              \
+            /* Set a bit to the last bit.                                     \
+             * Note that the following shifts bits first because pthread_t    \
+             * might use the last bit if, for example, pthread_t saves an ID  \
+             * starting from zero; overwriting the last bit can cause a       \
+             * conflict.  The last bit that is shifted out is less likely to  \
+             * be significant. */                                             \
+            id_tmp_ = (id_tmp_ << 1) | (uintptr_t)0x1;                        \
+        } else {                                                              \
+            id_tmp_ = (uintptr_t)self_thread_tmp_;                            \
+            /* If ID is that of Argobots, the last bit is not set because     \
+             * ABT_thread points to an aligned memory region.  Since          \
+             * ABT_thread is not modified, this ID can be directly used by    \
+             * MPL_thread_join() */                                           \
```

Comment:
Is there a possibility that the id returned by `MPL_thread_self()` is used for `MPL_thread_join()`? That'll be a mess and I would very much desire to remove that possibility -- maybe assertions in `MPL_thread_join`?

Separately, rather than commenting on a *assumed* behavior of an external library (argobot), how about *explicitly* clear the bit. That way the code is self explanatory and robust against argobot internal changes.

Reply:
Thanks for your comments.

> Is there a possibility that the id returned by `MPL_thread_self()` is used for `MPL_thread_join()`?

In my understanding, no. They should use the same ID rule, though. For example, it is reasonable to compare a thread ID obtained by `MPL_thread_self()` with an ID returned from `MPL_thread_create()` (and thus an ID used for `MPL_thread_join()`).

> That'll be a mess and I would very much desire to remove that possibility -- maybe assertions in MPL_thread_join?

Yes it is possible. Do you suggest that `MPL_thread_join()` have an assert to check if `MPL_thread_id_t` is "pthread_t"? 

If we want to generally prohibit using an ID obtained by MPL_thread_self()` for `MPL_thread_join()`, it needs an extra bit (and maybe I need to modify `MPL_thread_equal()`). I believe things get more complicated if we add a new bit to check whether `MPL_thread_id_t` is obtained by `MPL_thread_self()` or by `MPL_thread_create()`.

> Separately, rather than commenting on a assumed behavior of an external library (argobot), how about explicitly clear the bit. That way the code is self explanatory and robust against argobot internal changes.

It is self explanatory, but I don't think it is robust; the problem happens even if we clear those bits since clearing those bits just increase a chance of a conflict (see below) and in any case such a conflict will occur silently. I will add an assert to check the last bit of `ABT_thread`.

Note that, in any case, this pthread-or-not bit is not required for the current `MPL_thread` functionality though it would be beneficial for debugging and maybe some future extensions. The current mechanism of this PR works if `pthread_t` and `ABT_thread` are unique to each other. Even without a special bit, this should be reasonably true (`ABT_thread` points to a unique memory address. Reasonable implementation of `pthread_t` can be either an ID or a pointer to a memory address, so there should be no conflict between `ABT_thread` and `pthread_t`).

Reply:
I added an assert to check if the last bit of `ABT_thread` is surely zero